### ISSUE TITLE
"_global2.default.require is not a function" error on Node - Introduced in version 0.1.7

### ISSUE THREAD

**Issue by yoadsn** [Shopify/js-buy-sdk#153]
Hey guys,
I have upgraded to 0.1.7 yesterday and started experiencing the following error when the node application runs (I use Shopify Buy SDK both on the server and on the client):

```
d:\Development\Experiments\tmp_node_test\node_modules\shopify-buy\lib\isomorphic-fetch.js:16
  _global2.default.fetch = _global2.default.require('node-fetch');
                                            ^

TypeError: _global2.default.require is not a function
    at Object.<anonymous> (d:\Development\Experiments\tmp_node_test\node_modules\shopify-buy\lib\isomorphic-fetch.js:16:45)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (d:\Development\Experiments\tmp_node_test\node_modules\shopify-buy\lib\shopify.js:19:1)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
```

I have created a minimal possible scenario to make sure it is not something specific to my code:
1. empty folder - `npm init`
2. `npm install shopify-buy -- save`
3. created index.js with the following contents:
`var s = require("shopify-buy");`
4. Running `node index` produces the above error

Node version used: v5.6.0
Npm version used: 3.10.2

From my investigation, the `global.require` access used instead of globally using `require` (pre 0.1.7) is the culprit. For some reason, when running node as a REPL - `global.require` is defined. But when running it with a script - `global.require` is not defined.

So far, digging into node's repl and c++ code I could not find any reason for this difference. Also I am currently not able to test on node V4.2.2 which is the highest version your CI tests on.

@minasmart  Is there a reason you chose to not use the globally available "require" (after you confirmed it's a "node like environment") ?

**Comment by minasmart** [comment]
Hey! 

Thanks for investigating this so thoroughly. The problem with using `require` on its own is only that babel chokes on it. It's module compilers get confused when they see require conditionally nested. I'll look into this right away and see if there's a better workaround. I'm also open to recommendations if you think you have some insight into what the root of the problem may be.

Thanks Again!

**Comment by yoadsn** [comment]
I am sorry, but I'm still unable to build on windows even after you converted the scripts to node.
The environment is Windows 10, using Git Bash.
(I can see the source of the problem, if you think it's worth it I can open another issue with a repro and description..)

**Comment by yoadsn** [comment]
Another update, on my machine this issue occurs on node 4.4.2 as well.
I'm unclear as to how this passes CI.

**Comment by yomexzo** [comment]
If you refer to how the require issue passes, our CI tests currently run within a browser (PhantomJS). So that bit is never really experienced during the test. #131 will make that possible when completed.

re: build on windows, are you referring to the same issue described here @yoadsn?

**Comment by yoadsn** [comment]
@yomexzo Thanks, good to know how the CI works - I Only briefly looked at the setup yaml.
Regarding the build on windows, you have not referred to any issue number, but anyway it's an aside I don't want to pollute this thread with it.

**Comment by yoadsn** [comment]
@minasmart Hey, I eventually created a working build env under docker running linux (But really, I wish it was windows compatible so I could help more).
Having done that, I have observed the babel problem you described. Without investigating that I did try some naaive work around which worked for me:

`isomorphic-fetch.js`:

```
import global from './metal/global';
import isNodeLikeEnvironment from './metal/is-node-like-environment';

const fetch = global.fetch;

var _require = require;

if (!fetch && isNodeLikeEnvironment()) {
  global.fetch = _require('node-fetch');
  global.Response = global.fetch.Response;
}
```

Which transpiles to:
```
'use strict';

var _global = require('./metal/global');

var _global2 = _interopRequireDefault(_global);

var _isNodeLikeEnvironment = require('./metal/is-node-like-environment');

var _isNodeLikeEnvironment2 = _interopRequireDefault(_isNodeLikeEnvironment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var fetch = _global2.default.fetch;

var _require = require;

if (!fetch && (0, _isNodeLikeEnvironment2.default)()) {
  _global2.default.fetch = _require('node-fetch');
  _global2.default.Response = _global2.default.fetch.Response;
```

Since you already check if it's "node like" env, even if `_require` is undefined things would work fine in the browser. (I assume?)

**Comment by yomexzo** [comment]
That, definitely looks interesting. Could be the way to go. We'd see.

Good one @yoadsn.

**Comment by yoadsn** [comment]
@minasmart @yomexzo Sure - This got me curious as to why the transpilation fails this way (when require used as is).

When running your build pipeline I get the same error you got:
```
File: isomorphic-fetch.js
TypeError: isomorphic-fetch.js: Property right of AssignmentExpression expected node to be of a type ["Expression"] but instead got null
...
```
But when trying to install babel (`babel-cli@6.10.1`) and use that directly on the command line:
`babel --presets es2015 --plugins babel-plugin-transform-es2015-modules-commonjs,babel-plugin-transform-es2015-block-scoping,babel-plugin-transform-strict-mode ./src/isomorphic-fetch.js`
(Hopefully the right configuration as you use it in your pipeline)
It passes with no problem producing:

```
use strict';

var _global = require('./metal/global');

var _global2 = _interopRequireDefault(_global);

var _isNodeLikeEnvironment = require('./metal/is-node-like-environment');

var _isNodeLikeEnvironment2 = _interopRequireDefault(_isNodeLikeEnvironment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var fetch = _global2.default.fetch;

if (!fetch && (0, _isNodeLikeEnvironment2.default)()) {
  _global2.default.fetch = require('node-fetch');
  _global2.default.Response = _global2.default.fetch.Response;
```
as expected.

So I can only think the difference is in how the broccoli-babel (internally using babel 6.10 I believe) transpiler activates the transpilation or a version mismatch.

HTH

**Comment by minasmart** [comment]
It's specifically the AMD transpiler that chokes up. The commonjs one seems to pass.

As for working on windows: We've tested under powershell, and native bash on linux and mac. I'm not sure how git bash works on windows though. I'd certainly appreciate it if you could open an issue in regard to what's going on there.

I haven't had a ton of time to investigate a better way to grab the require var, but I'm glad your way seems to work. I should hopefully have something more concrete for you soon getting this to work.

**Comment by minasmart** [comment]
Relates to #164

**Comment by yomexzo** [comment]
fixed in v0.2.2
